content/blog/2020-12-09-slp/index.md

```diff
@@ -0,0 +1,184 @@
+   - The statements do the same operation on independent data.
+   - It is beneficial to execute these at once rather than as multiple 
+     instructions. If the data is already packed, then there's no additional 
+     cost. However, if the data is not packed we have to take into account the 
+     cost of moving them into the SIMD register.
+   - The positions of the values in the SIMD register is consistent with the 
+     previous packed statements. This is what the paper calls alignment.
+   - To simplify the search space for the next part, we 
+     only allow a statement to occupy the beginning or end instruction in a 
+     pack list once. This way, we don't have to choose between multiple options 
+     later down the line.
+   ![phase2](phase2.png)
+3. Now that we have pairs of instructions that can be executed by a single SIMD
+   instruction without any issues, we can combine multiple of these pairs into 
+   an even wider SIMD instruction. Taking advantage of the last bullet above, 
+   we only combine two lists if the beginning and end instruction is the same.
+   ![phase3](phase3.png)
+4. The last part is to actually generate the concrete SIMD instructions.
+   Since statement groups could depend on data defined by a previous one, we
+   need to generate the instructions in a topological ordering of the data
+   dependency graph. The original instruction order works so the scheduling 
+   phase generates instructions in the order of the original basic block. To 
+   deal with cycles, we simply revert the pack and use regular instructions 
+   instead.
+   ![phase4](phase4.png)
+
+This routine will take in a single basic block and then convert it to a basic
+block with SIMD instructions.
+
+By simply unrolling loops, we can apply this to loops to get a vector parallel 
+version.
+
+## Performance Comparison
+
+The authors chose to use a benchmark suite that consists of multimedia and 
+scientific programs. These process heavy amounts of data, e.x. 
+matrix multiplication or color conversion. Their experimental platform was a 
+Motorola processor with an AltiVec co-processor, i.e. all vector instructions 
+work on the coprocessor and all sequential operations on the main unit. Any 
+movement between the vector registers and sequential registers must pass 
+through memory.
+
+They don't really compare with any other baseline implementations. They mention 
+previous versions of this in their related work section but it would have been 
+nice to see how many opportunities SLP can exploit over the loop transformation 
+version. However, they do show which opportunities come from vectorizable 
+(loop based) opportunities versus just SLP, which still shows large improvement,
+just not as large.
+
+One claim that they put forth (albeit a reasonable one) is that their approach
+is more robust than the previous loop transformation techniques. It would be
+have been good to see this experimentally verified in their chosen benchmarks.
```

Comment:
Agreed. IMO they're complementary techniques. For example, LLVM has both a loop vectorizer and an SLP vectorizer. https://llvm.org/docs/Vectorizers.html